### ISSUE TITLE
Bug: devtools crash when useState and useEffect used twice

### ISSUE THREAD

**Issue by avkonst** [facebook/react#20613]
React version: 17.0.1

## Steps To Reproduce

1. clone repo: https://github.com/avkonst/react-devtools-crash-demo
2. yarn start
3. open the browser of the started app
4. open devtools
5. click Counter component in the components tab of the development tools
6. Watch it not loading anything and unhandled exception dumped to the console log

Link to code example: https://github.com/avkonst/react-devtools-crash-demo

## The current behavior

Crash when the following hook is used twice in a component:

```tsx
export function useState() {
    React.useState(0);
    React.useEffect(() => {
        return () => {}
    });
}
```

## The expected behavior

No crash and the development tools show hooks used.

**Comment by rickhanlonii** [comment]
@avkonst thanks for submitting, continuing the discussion from https://github.com/facebook/react/issues/20609.

It's very common for stack traces to point to React internals where the issue is sourced in user code. Since you're the expert of the way your library works, can you strip down the example to use vanilla React?

**Comment by avkonst** [comment]
Have you seen the code in the demo repo? I stripped down the example to just that:
```
export function useState() {
    React.useState(0);
    React.useEffect(() => {
        return () => {}
    });
}
```
This is vanilla react and this hook causes the crash if used twice in a component. I can not make it smaller than this. Is it good enough?

**Comment by avkonst** [comment]
Here is the complete reproducer in 18 lines of vanilla react code:

```tsx
import React from 'react';

function useState() {
    React.useState(0);
    React.useEffect(() => {
        return () => {}
    });
}

function Counter() {
	useState();
	React.useState(0);
	return (
		<div>Open React Dev Tools Components panel, click on Counter component and observe the crash in the logging console.</div>
	);
}

export default Counter;
```

**Comment by avkonst** [comment]
Committed this sample to the demo repo. Is it good enough?

**Comment by avkonst** [comment]
I have updated the original description too. There is nothing left from Hookstate in the above 16 lines of vanilla react code.

**Comment by avkonst** [comment]
Committed the same code sample to the demo repository now.

**Comment by dai-shi** [comment]
Issue reproduced, and it's a naming conflict? Renaming the custom hook to `useMyState` seems to work.
https://codesandbox.io/s/ecstatic-shockley-g1usv

**Comment by avkonst** [comment]
Maybe naming. But devtools should not influence how I name my functions, right? Compiler is happy with this name, so why not...

**Comment by rickhanlonii** [comment]
@avkonst thanks a ton for following up and reducing the code down.

The issue is that DevTools errors when a custom hook is named `useState`. If you rename the custom hook in your example to `useStateCustom`, DevTools does not error.

**Comment by avkonst** [comment]
Sure, but I can not rename the hook as it is the interface of the public library and I could not find better name originally. Hookstate.useState is in the different module and naming space than React.useState. It does not cause any conflict for the compiler and bundler, why should it for devtools. Hookstate.useState is the supercharged version of React.useState, hence named identically as it is nearly plug and play replacement.

**Comment by rickhanlonii** [comment]
Yeah, I agree, it probably shouldn't error but I don't have a much context as @bvaughn so let's wait to see what he says.

In the meantime, I've created a failing test here https://github.com/facebook/react/pull/20614.

Thanks again for reporting the issue and stripping down the issue so I could debug it quickly.

**Comment by rickhanlonii** [comment]
And thanks @dai-shi for debugging this as well, nice find!

**Comment by bvaughn** [comment]
Thanks for the bug report and concise repo. I'll take a look this morning.

**Comment by bvaughn** [comment]
I can reproduce this failure in a unit test btw:
```js
it('should gracefully handle custom hooks that have the same name as primitive hooks', async done => {
  let inspectedElement = null;
  function Suspender({target}) {
    inspectedElement = useInspectedElement(target);
    return null;
  }

  const container = document.createElement('div');

  function useState() {
    React.useState(0);
    React.useEffect(() => {});
  }

  function HooksNamingConflict() {
    useState();
    React.useState(0);
    return null;
  }

  await utils.actAsync(() =>
    ReactDOM.render(<HooksNamingConflict />, container),
  );

  const id = ((store.getElementIDAtIndex(0): any): number);
  await utils.actAsync(
    () =>
      TestRenderer.create(
        <Contexts
          defaultSelectedElementID={id}
          defaultSelectedElementIndex={0}>
          <React.Suspense fallback={null}>
            <Suspender target={id} />
          </React.Suspense>
        </Contexts>,
      ),
    false,
  );
  expect(inspectedElement.hooks).toMatchInlineSnapshot(`
    Array [
      Object {
        "id": 0,
        "isStateEditable": true,
        "name": "State",
        "subHooks": Array [],
        "value": 1,
      },
    ]
  `);

  done();
});
```

Seems like a few things are required:
* Custom hook with the same name as built-in hook
* Use the same name built-in hook _and_ at least one other built-in hook inside of the custom hook
* Call both the custom hook and at least one other built-in hook in a React component

**Comment by rickhanlonii** [comment]
@bvaughn in your smaller version, I don't think you need the useEffect

**Comment by bvaughn** [comment]
Seems like (maybe among other things) this check is not great:
https://github.com/facebook/react/blob/master/packages/react-debug-tools/src/ReactDebugHooks.js#L538-L539

**Comment by bvaughn** [comment]
> @bvaughn in your smaller version, I don't think you need the useEffect

You do.

**Comment by bvaughn** [comment]
I think the core of this problem is here:
https://github.com/facebook/react/blob/2e948e0d91360e4763594854be84347737e4b1f6/packages/react-debug-tools/src/ReactDebugHooks.js#L423-L444

This code is seeing the custom `useState` wrapper and mistaking it for a wrapper added by a packager around the native hook. As a result it's bumping the index higher than it should.

Not sure yet how to fix _this_ case without breaking _that_ one though (or support for composite hooks like `useMutableSource`). Need to think about it more.

**Comment by bvaughn** [comment]
I need to sit this task down for now because I have something more pressing to work on. If anyone else feels like picking it up, that would be great. We have pretty good test coverage in `react-debug-tools`.

**Comment by bvaughn** [comment]
Note that I think we could prevent the runtime error by adding a check before pushing to `levelChildren`:
```js
if (levelChildren !== undefined) {
  levelChildren.push({
    id,
    isStateEditable,
    name: primitive,
    value: hook.value,
    subHooks: [],
  });
}
```

But this would have two downsides:
* It would mask the underlying error.
* It would hide some of the hooks information.

Consider the following component example:
```js
function useState() {
  React.useState(0);
  React.useEffect(() => {});
}

function HooksNamingConflict() {
  useState();
  React.useMemo(() => 0, []);
  return null;
}
```

We should see a hooks tree of:
* State (custom)
  * State (actual)
  * Effect
* Memo

With the above fix we would only see a reported tree of:
* State (actual)
* Memo

Stepping up the stack too far in `findPrimitiveIndex` would cause us to miss the custom `useState` as well as any other primitive hooks it calls inside of it.

**Comment by avkonst** [comment]
Brian, would it be possible to apply and merge the solution you proposed in
the latest comment above? The downside you mentioned is actually a good
feature for hookstate support because it would hide internal hooks used by
hookstate and would show only the actual state hook. I think it is the best
outcome we could get really...

**Comment by bvaughn** [comment]
That's not the way custom hook support is supposed to work in DevTools, and I don't think it would be what you wanted anyway (e.g. Effect hook would be invisible in this scenario).

**Comment by avkonst** [comment]
Brian, yes, I want the effect hook invisible when it is inside of custom
hook named useState. I understand you proposal does not have the downside
for all other custom hooks, right?

**Comment by bvaughn** [comment]
It just happens to be the effect hook because of the way this example code is structured. It could be other arbitrary hooks that were shown or hidden if the code were structured differently. That's not a design that I think we would be able to land.

**Comment by avkonst** [comment]
Are you planning to return to this back again? We need more ideas :) I
certainly have got no idea and zero knowledge of this code...

**Comment by bvaughn** [comment]
I wrote [this comment](https://github.com/facebook/react/issues/20613#issuecomment-763755563) only a few moments ago.
> I need to sit this task down for now because I have something more pressing to work on. If anyone else feels like picking it up, that would be great. We have pretty good test coverage in `react-debug-tools`.

**Comment by avkonst** [comment]
Guys, any plans or ideas when/how to progress this?

**Comment by avkonst** [comment]
Sorry, I missed the pull request progress. Seems like React team is getting close to resolve it. Great. Very much appreciated.

**Comment by avkonst** [comment]
Hi guys. What is the plan for this issue? There was some pull request, but it is unclear where it is gone...

**Comment by avkonst** [comment]
Could you please give an update? Has it been fixed?

**Comment by rickhanlonii** [comment]
This should be fixed in https://github.com/facebook/react/pull/28413
